template-coq/src/pluginCore.ml

```diff
@@ -1,57 +1,167 @@
 (* this file is the interface that extracted plugins will use.
  *)
 
-type ident   (* Template.Ast.ident *)
-type kername (* Template.Ast.kername *)
-type reductionStrategy (* Template.TemplateMonad.Common.reductionStrategy *)
-type global_reference (* Template.Ast.global_reference *)
+type ident   = Names.Id.t (* Template.Ast.ident *)
+type kername = Names.KerName.t (* Template.Ast.kername *)
+type reduction_strategy = Redexpr.red_expr (* Template.TemplateMonad.Common.reductionStrategy *)
+type global_reference = Globnames.global_reference (* Template.Ast.global_reference *)
 type term = Constr.t  (* Ast.term *)
+type mutual_inductive_body = Declarations.mutual_inductive_body (* Template.Ast.mutual_inductive_body *)
+type constant_entry = Declarations.constant_body (* Template.Ast.constant_entry *)
+type mutual_inductive_entry = Entries.mutual_inductive_entry (* Template.Ast.mutual_inductive_entry *)
 
-type 'a tm = 'a
+let default_flags = Redops.make_red_flag Genredexpr.[FBeta;FMatch;FFix;FCofix;FZeta;FDeltaBut []]
+let rs_cbv = Genredexpr.Cbv default_flags
+let rs_cbn = Genredexpr.Cbn default_flags
+let rs_hnf = Genredexpr.Hnf
+let rs_all = Genredexpr.Cbv Redops.all_flags
+let rs_lazy = Genredexpr.Cbv Redops.all_flags
+
+(* question(gmm): maybe this should just take a global_reference? *)
+let rs_unfold (env : Environ.env) (kn : kername) =
+  (* note(gmm): this shouldn't be necessary *)
+  let ident = Names.Id.of_string (Names.KerName.to_string kn) in
+  try
+    let gr = (Nametab.global (CAst.make (Libnames.Qualid (Libnames.qualid_of_ident ident)))) in
+    Genredexpr.Unfold [Locus.AllOccurrences,
+                       Tacred.evaluable_of_global_reference env gr]
+  with
+  | _ -> CErrors.user_err Pp.(str "Constant not found or not a constant: " ++ str (Names.Id.to_string ident))
+
+type 'a tm =
+  { run_tm : Environ.env -> Evd.evar_map ->
+      (Environ.env -> Evd.evar_map -> 'a -> unit) -> (unit -> unit) -> unit }
```

Comment:
Originally I made this polymorphic (standard CPS monad), but that doesn't work due to `tmLemma` (mentioned in the comment). This means that I could drop the record.